File: android-java/app/src/main/java/io/agora/convoai/example/voiceassistant/ui/common/VoiceWaveView.java
package io.agora.convoai.example.voiceassistant.ui.common;

import android.animation.ValueAnimator;
import android.content.Context;
import android.content.res.Resources;
import android.graphics.Canvas;
import android.graphics.Paint;
import android.graphics.RectF;
import android.util.AttributeSet;
import android.util.TypedValue;
import android.view.View;
import android.view.animation.LinearInterpolator;
import androidx.core.content.ContextCompat;
import io.agora.convoai.example.voiceassistant.java.R;
import java.util.Random;

/**
 * Agent Speaking Indicator
 * Four bars, wave-like random animation
 */
public class VoiceWaveView extends View {

    private static final int BAR_COUNT = 4;
    private static final float BAR_WIDTH = dpToPx(5);
    private static final float BAR_SPACING = dpToPx(6);
    private static final float BAR_CORNER_RADIUS = dpToPx(3);
    private static final float BAR_HEIGHT_MIN = dpToPx(5);
    private static final float BAR_HEIGHT_MAX = dpToPx(12);
    private static final long ANIMATION_DURATION = 1400L; // ms, one wave cycle, slower for smoothness
    private static final float PHASE_DRIFT_PER_FRAME = 0.018f; // phase drift per frame for flowing effect
    private static final float JITTER_AMPLITUDE = dpToPx(0.5f); // smaller jitter for less abruptness

    private final Paint paint;
    private final float[] barHeights = new float[BAR_COUNT];
    private ValueAnimator animator;
    private float animationProgress = 0f;
    private final float[] phaseOffsets = new float[BAR_COUNT];
    private final Random random = new Random();

    private static float dpToPx(float dp) {
        return TypedValue.applyDimension(
            TypedValue.COMPLEX_UNIT_DIP,
            dp,
            Resources.getSystem().getDisplayMetrics()
        );
    }

    public VoiceWaveView(Context context) {
        this(context, null);
    }

    public VoiceWaveView(Context context, AttributeSet attrs) {
        this(context, attrs, 0);
    }

    public VoiceWaveView(Context context, AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        
        paint = new Paint(Paint.ANTI_ALIAS_FLAG);
        paint.setColor(ContextCompat.getColor(context, R.color.white));
        paint.setStyle(Paint.Style.FILL);
        
        // Initialize bar heights to minimum
        for (int i = 0; i < BAR_COUNT; i++) {
            barHeights[i] = BAR_HEIGHT_MIN;
            phaseOffsets[i] = random.nextFloat() * 2f * (float) Math.PI;
        }
    }

    /**
     * Start wave animation
     */
    public void startAnimation() {
        if (animator != null && animator.isRunning()) {
            return;
        }
        
        animator = ValueAnimator.ofFloat(0f, 1f);
        animator.setDuration(ANIMATION_DURATION);
        animator.setRepeatCount(ValueAnimator.INFINITE);
        animator.setRepeatMode(ValueAnimator.RESTART);
        animator.setInterpolator(new LinearInterpolator());
        animator.addUpdateListener(animation -> {
            animationProgress = (Float) animation.getAnimatedValue();
            updateBarHeights();
            invalidate();
        });
        animator.start();
    }

    /**
     * Stop animation and reset bars to min height
     */
    public void stopAnimation() {
        if (animator != null) {
            animator.cancel();
            animator = null;
        }
        for (int i = 0; i < BAR_COUNT; i++) {
            barHeights[i] = BAR_HEIGHT_MIN;
        }
        invalidate();
    }

    private void updateBarHeights() {
        for (int i = 0; i < BAR_COUNT; i++) {
            phaseOffsets[i] = (phaseOffsets[i] + PHASE_DRIFT_PER_FRAME) % (2f * (float) Math.PI);
            double wave = Math.sin(2 * Math.PI * (animationProgress + phaseOffsets[i]));
            float base = BAR_HEIGHT_MIN + (BAR_HEIGHT_MAX - BAR_HEIGHT_MIN) * ((float) wave + 1f) / 2f;
            float jitter = random.nextFloat() * 2f - 1f; // [-1, 1]
            barHeights[i] = Math.max(BAR_HEIGHT_MIN, Math.min(BAR_HEIGHT_MAX, base + jitter * JITTER_AMPLITUDE));
        }
    }

    @Override
    protected void onDraw(Canvas canvas) {
        super.onDraw(canvas);
        float centerX = getWidth() / 2f;
        float centerY = getHeight() / 2f;
        float totalWidth = BAR_COUNT * BAR_WIDTH + (BAR_COUNT - 1) * BAR_SPACING;
        float startX = centerX - totalWidth / 2f;
        
        for (int i = 0; i < BAR_COUNT; i++) {
            float barX = startX + i * (BAR_WIDTH + BAR_SPACING);
            float barTop = centerY - barHeights[i] / 2f;
            float barBottom = centerY + barHeights[i] / 2f;
            RectF rect = new RectF(barX, barTop, barX + BAR_WIDTH, barBottom);
            canvas.drawRoundRect(rect, BAR_CORNER_RADIUS, BAR_CORNER_RADIUS, paint);
        }
    }

    @Override
    protected void onMeasure(int widthMeasureSpec, int heightMeasureSpec) {
        float totalWidth = BAR_COUNT * BAR_WIDTH + (BAR_COUNT - 1) * BAR_SPACING;
        int desiredWidth = (int) totalWidth;
        int desiredHeight = (int) BAR_HEIGHT_MAX;
        int width = resolveSize(desiredWidth, widthMeasureSpec);
        int height = resolveSize(desiredHeight, heightMeasureSpec);
        setMeasuredDimension(width, height);
    }

    @Override
    protected void onDetachedFromWindow() {
        super.onDetachedFromWindow();
        if (animator != null) {
            animator.cancel();
            animator = null;
        }
    }
}

File: android-compose/app/src/main/java/io/agora/convoai/example/compose/voiceassistant/ui/VoiceWaveView.kt
package io.agora.convoai.example.compose.voiceassistant.ui

import androidx.compose.animation.core.*
import androidx.compose.foundation.Canvas
import androidx.compose.foundation.layout.size
import androidx.compose.runtime.*
import androidx.compose.ui.Modifier
import androidx.compose.ui.geometry.CornerRadius
import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.geometry.Size
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.tooling.preview.Preview
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.times
import kotlin.math.PI
import kotlin.math.sin
import kotlin.random.Random

/**
 * Agent Speaking Indicator
 * Four bars, wave-like random animation
 * @param scale Scale factor for all dimensions (default: 1.0f)
 */
@Composable
@Preview
fun VoiceWaveView(
    modifier: Modifier = Modifier,
    isAnimating: Boolean = false,
    color: Color = Color.White,
    scale: Float = 1.0f
) {
    val barCount = 4
    val baseBarWidth = 6.dp
    val baseBarSpacing = 8.dp
    val baseBarCornerRadius = 3.dp
    val baseBarHeightMin = 6.dp
    val baseBarHeightMax = 15.dp
    
    val barWidth = (baseBarWidth * scale)
    val barSpacing = (baseBarSpacing * scale)
    val barCornerRadius = (baseBarCornerRadius * scale)
    val barHeightMin = (baseBarHeightMin * scale)
    val barHeightMax = (baseBarHeightMax * scale)

    val animationDuration = 1400 // ms
    val phaseDriftPerFrame = 0.018f

    // Animation state
    val infiniteTransition = rememberInfiniteTransition(label = "voice_wave")
    val animationProgress by infiniteTransition.animateFloat(
        initialValue = 0f,
        targetValue = 1f,
        animationSpec = infiniteRepeatable(
            animation = tween(animationDuration, easing = LinearEasing),
            repeatMode = RepeatMode.Restart
        ),
        label = "progress"
    )

    // Phase offsets for each bar
    val phaseOffsets = remember { FloatArray(barCount) { Random.nextFloat() * 2f * PI.toFloat() } }
    
    // Bar heights (stored as normalized values 0.0-1.0, will be converted to pixels in Canvas)
    val barHeights = remember { mutableStateListOf<Float>().apply {
        repeat(barCount) { add(0f) } // Start at minimum (0.0 = min height)
    }}

    // Update bar heights when animating
    LaunchedEffect(isAnimating, animationProgress) {
        if (isAnimating) {
            for (i in 0 until barCount) {
                // Update phase offset with drift
                phaseOffsets[i] = (phaseOffsets[i] + phaseDriftPerFrame) % (2f * PI.toFloat())
                
                // Calculate wave height (normalized 0.0-1.0)
                val wave = sin(2 * PI * (animationProgress + phaseOffsets[i])).toFloat()
                val normalized = ((wave + 1f) / 2f) // Convert from [-1, 1] to [0, 1]
                
                // Add random jitter
                val jitter = Random.nextFloat() * 2f - 1f // [-1, 1]
                val jitterAmplitude = 0.1f // 10% jitter
                val height = (normalized + jitter * jitterAmplitude).coerceIn(0f, 1f)
                
                barHeights[i] = height
            }
        } else {
            // Reset to min height when not animating
            for (i in 0 until barCount) {
                barHeights[i] = 0f
            }
        }
    }

    Canvas(
        modifier = modifier
            .size(
                width = (barCount * barWidth + (barCount - 1) * barSpacing),
                height = barHeightMax * 1.2f // Add extra height to prevent clipping and ensure proper aspect ratio
            )
    ) {
        val centerY = size.height / 2f
        val barWidthPx = barWidth.toPx()
        val barSpacingPx = barSpacing.toPx()
        val barCornerRadiusPx = barCornerRadius.toPx()
        val totalWidth = barCount * barWidthPx + (barCount - 1) * barSpacingPx
        val startX = (size.width - totalWidth) / 2f

        // Convert Dp to pixels for height calculations
        val barHeightMinPx = barHeightMin.toPx()
        val barHeightMaxPx = barHeightMax.toPx()
        
        for (i in 0 until barCount) {
            val barX = startX + i * (barWidthPx + barSpacingPx)
            // Convert normalized height (0.0-1.0) to actual pixel height
            val normalizedHeight = barHeights[i]
            val currentBarHeight = barHeightMinPx + (barHeightMaxPx - barHeightMinPx) * normalizedHeight
            val barTop = centerY - currentBarHeight / 2f

            drawRoundRect(
                color = color,
                topLeft = Offset(barX, barTop),
                size = Size(barWidthPx, currentBarHeight),
                cornerRadius = CornerRadius(barCornerRadiusPx, barCornerRadiusPx)
            )
        }
    }
}

